server: use "read-only" Gossip for tenants

tbg opened the pull request:
We were previously using the Gossip instance of the TestServer against
which the tenant was initialized.

This commit trims the dependency further by initializing its own Gossip
instance which is never written to (i.e. `AddInfo` is never called) and
which does not accept incoming connections.

As a reminder, the remaining problematic uses of Gossip as of this
commit are:

- making a `nodeDialer` (for `DistSender`), tracked in:
  #47909
- access to the system config:
  - `(schemaChangeGCResumer).Resume`, tracked:
    #49691
  - `database.Cache`, tracked:
    #49692
- `(physicalplan).spanResolver` (for replica oracle).
  This is likely not a blocker as we can use a "dumber" oracle in this case;
  the oracle is used for distsql physical planning of which tenants will
  do none. Tracked in: #48432

cc @ajwerner 

Release note: None

tbg committed:
server: add Make{SQL,Base}Config

Release note: None

tbg committed:
server: introduce KVConfig

Release note: None

tbg committed:
server: fold tenantID into SQLConfig

Release note: None

tbg committed:
server: makeTestConfig -> makeTest{SQL,Base,KV}Config

Release note: None

tbg committed:
server: use "read-only" Gossip for tenants

We were previously using the Gossip instance of the TestServer against
which the tenant was initialized.

This commit trims the dependency further by initializing its own Gossip
instance which is never written to (i.e. `AddInfo` is never called) and
which does not accept incoming connections.

As a reminder, the remaining problematic uses of Gossip as of this
commit are:

- making a `nodeDialer` (for `DistSender`), tracked in:
  https://github.com/cockroachdb/cockroach/issues/47909
- access to the system config:
  - `(schemaChangeGCResumer).Resume`, tracked:
    https://github.com/cockroachdb/cockroach/issues/49691
  - `database.Cache`, tracked:
    https://github.com/cockroachdb/cockroach/issues/49692
- `(physicalplan).spanResolver` (for replica oracle).
  This is likely not a blocker as we can use a "dumber" oracle in this case;
  the oracle is used for distsql physical planning of which tenants will
  do none. Tracked in: https://github.com/cockroachdb/cockroach/issues/48432

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49693)
<!-- Reviewable:end -->

tbg committed:
server: make cli-ready startTenant method()

The goal is to turn `testSQLServerArgs` into something that can be
re-used to introduce a cli command for starting a SQL tenant. This
commit takes a step in that direction by lifting some testing-specific
inputs up.

Release note: None

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-:-M8jBWBd1gbHXlt5wO2X:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/server/config.go, line 292 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-M8j7BMN-liGC1mOApd5:-M8j7BMN-liGC1mOApd6:b-yc2383) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1ff884fb44c0b593281483e0c51d7f633c683e/pkg/server/config.go#L292)):*
> ```Go
> 		},
> 	}
> 	kvCfg.SetDefaults()
> ```

nit: I find `kvCfg.RaftConfig.SetDefaults` easier to read, up to you
___
*[pkg/server/testserver.go, line 636 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-M8jAuUQ9YFIKnMFWQCH:-M8jAuUQ9YFIKnMFWQCI:b67qdng) ([raw file](https://github.com/cockroachdb/cockroach/blob/e487d70be7c299a0458e8dd6496d67ffa2a00b3c/pkg/server/testserver.go#L636)):*
> ```Go
> 		}
> 	}
> 	ctx := context.Background()
> ```

nit: pass this context in?
___
*[pkg/server/testserver.go, line 580 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-M8jBJEG5CZwDI4mxY3M:-M8jBJEG5CZwDI4mxY3N:b2hjxpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a27edcc5ff7e728cd64e691f853e3e581473e0d/pkg/server/testserver.go#L580)):*
> ```Go
> 			recorder:     dummyRecorder,
> 			isMeta1Leaseholder: func(timestamp hlc.Timestamp) (bool, error) {
> 				return false, errors.New("isMeta1Leaseholder is not available to tenants")
> ```

`s/to tenants/to non-system tenants`?


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 4 of 4 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, 2 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-:-M8kMOMLCweCY2YYaZrP:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/server/config.go, line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-M8kKQiqAqJR2myIF9RJ:-M8kKQiqAqJR2myIF9RK:b-bi4869) ([raw file](https://github.com/cockroachdb/cockroach/blob/74b8bc043f9f3ac2016484561f9aca67110826a8/pkg/server/config.go#L145)):*
> ```Go
> 	baseCfg := BaseConfig{
> 		AmbientCtx:        log.AmbientContext{Tracer: st.Tracer},
> 		Config:            new(base.Config),
> ```

nit: any reason to embed a pointer?


<!-- Sent from Reviewable.io -->

tbg committed:
testserver: massage mock error messages

Release note: None

tbg committed:
server: pass ctx to startTenant

Release note: None

tbg reviewed (commented):
TFTRs!

Comments addressed.

bors r=asubiotto,nvanbenschoten

> Dismissed @asubiotto from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49693#-:-M8okUn56sjCOvKtRbGr:b-hxpzit)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

tbg commented:
Embedding `base.Config` by value broke the build pretty badly. I'll just remove that commit for now, we're likely now taking a copy where we previously changed the original. Not worth holding this PR up for but it's on my TODO list.

bors r=asubiotto,nvanbenschoten
